internal/expr/parser.go

```diff
@@ -716,30 +735,139 @@ func (p *Parser) parseOutputExpr() (*outputExpr, bool, error) {
 	}
 
-	// Case 1: Slice range, "Type[:]".
 	if st, ok, err := p.parseSliceAccessor(); err != nil {
+		cp.restore()
 		return nil, false, err
 	} else if ok {
 		return &sliceInputExpr{sliceTypeName: st, raw: p.input[cp.pos:p.pos]}, true, nil
 	}
 
-	// Case 2: Struct or map, "Type.something".
-	if ma, ok, err := p.parseTypeAndMember(); ok {
-		if ma.memberName == "*" {
-			return nil, false, errorAt(fmt.Errorf("asterisk not allowed in input expression %q", "$"+ma.String()), cp.lineNum, cp.colNum(), p.input)
+	cp.restore()
+	return nil, false, nil
+}
+
+// parseMemberInputExpr parses an input expression of the form "$Type.member".
+func (p *Parser) parseMemberInputExpr() (*memberInputExpr, bool, error) {
+	cp := p.save()
+	ma, ok, err := p.parseInputMemberAccessor()
+	if !ok {
+		cp.restore()
+		return nil, false, err
+	}
+	if ma.memberName == "*" {
+		cp.restore()
+		return nil, false, errorAt(fmt.Errorf("invalid asterisk input placement %q", "$"+ma.String()), cp.lineNum, cp.colNum(), p.input)
+	}
+	return &memberInputExpr{ma: ma, raw: p.input[cp.pos:p.pos]}, true, nil
+}
+
+// parseAsteriskInputExpr parses an INSERT statement input expression where
+// SQLair generates the columns.
+// It is of the form "(*) VALUES ($Type.*, $Type.member,...)".
+func (p *Parser) parseAsteriskInputExpr() (*asteriskInputExpr, bool, error) {
+	cp := p.save()
+
+	parenCol := p.pos + 1
+	if !p.skipString("(*)") {
+		return nil, false, nil
+	}
+	p.skipBlanks()
+	if !p.skipString("VALUES") {
+		cp.restore()
+		return nil, false, nil
+	}
+	p.skipBlanks()
+
+	sources, ok, err := parseList(p, (*Parser).parseInputMemberAccessor)
+	if err != nil {
+		cp.restore()
+		return nil, false, err
+	} else if !ok {
+		// Check for types with missing parentheses.
+		if _, ok, _ := p.parseInputMemberAccessor(); ok {
+			err = errorAt(fmt.Errorf(`missing parentheses around types after "VALUES"`), p.lineNum, parenCol, p.input)
+		}
+		cp.restore()
+		return nil, false, err
+	}
+
+	return &asteriskInputExpr{sources: sources, raw: p.input[cp.pos:p.pos]}, true, nil
+}
+
+// parseColumnsInputExpr parses an INSERT statement input expression where the
+// user specifies the columns to insert from a single type.
+// It is of the form "(col1, col2,...) VALUES ($Type.*)".
+func (p *Parser) parseColumnsInputExpr() (*columnsInputExpr, bool, error) {
+	cp := p.save()
+
+	columns, paren, ok := p.parseColumns()
+	if !(ok && paren) {
+		cp.restore()
+		return nil, false, nil
+	}
+
+	p.skipBlanks()
+	if !p.skipString("VALUES") {
+		cp.restore()
+		return nil, false, nil
+	}
+	p.skipBlanks()
+
+	parenCol := p.colNum()
+	// Ignore errors and leave them to be handled by
+	// parseMemberInputExpr later.
+	sources, ok, _ := parseList(p, (*Parser).parseInputMemberAccessor)
+	if !ok {
+		// Check for types with missing parentheses.
+		if _, ok, _ := p.parseTypeAndMember(); ok {
+			return nil, false, errorAt(fmt.Errorf(`missing parentheses around types after "VALUES"`), p.lineNum, parenCol, p.input)
```

Comment:
Good idea! Do we want to `cp.restore()` here, I think that we need some general advice, like do not restore if you are going to emit an error but do in all other cases.

Reply:
The policy is to restore every-time, even when emitting an error, just cuz the error could be ignored by the caller.